pkg/controllers/tokentimeoutsync/tokentimeoutsync.go

```diff
@@ -0,0 +1,179 @@
+	configv1lister "github.com/openshift/client-go/config/listers/config/v1"
+	oauthv1client "github.com/openshift/client-go/oauth/clientset/versioned/typed/oauth/v1"
+	oauthinformer "github.com/openshift/client-go/oauth/informers/externalversions"
+	oauthv1lister "github.com/openshift/client-go/oauth/listers/oauth/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+type tokenTimeoutSyncController struct {
+	oauthConfigLister configv1lister.OAuthLister
+	oauthClientLister oauthv1lister.OAuthClientLister
+	accessTokens      oauthv1client.OAuthAccessTokenInterface
+
+	globalTimeout  time.Duration
+	clientTimeouts map[string]*time.Duration
+}
+
+// knownConditionNames lists all condition types used by this controller.
+// These conditions are operated and defaulted by this controller.
+// Any new condition used by this controller sync() loop should be listed here.
+var knownConditionNames = sets.NewString(
+	"OAuthConfigGetFailed",
+	"OAuthClientListFailed",
+	"TokenListFailed",
+)
+
+func NewTokenTimeoutSyncController(
+	accessTokenClient oauthv1client.OAuthAccessTokenInterface,
+	oauthInformers oauthinformer.SharedInformerFactory,
+	configInformers configinformer.SharedInformerFactory,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &tokenTimeoutSyncController{
+		accessTokens:      accessTokenClient,
+		oauthConfigLister: configInformers.Config().V1().OAuths().Lister(),
+		oauthClientLister: oauthInformers.Oauth().V1().OAuthClients().Lister(),
+
+		globalTimeout:  -1,
+		clientTimeouts: map[string]*time.Duration{},
+	}
+
+	return factory.New().ResyncEvery(30*time.Second).WithInformers(
+		oauthInformers.Oauth().V1().OAuthClients().Informer(),
+		configInformers.Config().V1().OAuths().Informer(),
+	).WithSync(c.sync).ToController("TokenTimeoutSyncController", recorder.WithComponentSuffix("wellknown-ready-controller"))
+}
+
+func (c *tokenTimeoutSyncController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	var globalChanged bool
+
+	oauthConfig, err := c.oauthConfigLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	if oauthConfig.Spec.TokenConfig.AccessTokenInactivityTimeout != nil {
+		if newTimeout := oauthConfig.Spec.TokenConfig.AccessTokenInactivityTimeout.Duration; c.globalTimeout != newTimeout { // don't do anything on first run
+			// only update if this is not the first run and the new new timeout is lower than the previous
+			globalChanged = c.globalTimeout != -1 && ((c.globalTimeout > 0 && newTimeout < c.globalTimeout) || (c.globalTimeout == 0 && newTimeout > 0))
+			c.globalTimeout = newTimeout
+		}
+	} else {
+		globalChanged = c.globalTimeout != 0
+		c.globalTimeout = 0
+	}
+
+	tokenList, err := c.accessTokens.List(ctx, metav1.ListOptions{LabelSelector: labels.Everything().String()})
+	if err != nil {
+		return err
+	}
+
+	for _, t := range tokenList.Items {
```

Comment:
why not skip this if `!globalChanged`?

Reply:
clients could have changed their timeouts